### PR TITLE
Reflection updates

### PR DESCRIPTION
**Pull request by nathanielhourt**
## Change Description
This pull request is as much an effort to open a conversation as it is to submit a patch. All of these changes are negotiable.

I found myself working on my reflection system in client code to Leap/CDT and as a result, was looking at Leap's reflection. It seemed to me that the reflection was partially implemented in many places and generally in need of some tuneup, so I have made several improvements, and if these are agreeable, I can make more.

The goal for me is to make my reflection and Leap's compatible, so when I define reflection using my infrastructure, I get Leap's reflection records defined too, but along the way, I hope to zhuzh up Leap's reflection a bit.

So let's talk: do the changes make sense on your end? I realize they alter a lot about how the reflection was implemented, so if some of the things I removed were important, help me understand why and I'll see if I can work them back in. :-)

## API Changes
- [x] API Changes

This PReq specifically does *not* alter the main reflection interface, although if [an alteration](https://github.com/AntelopeIO/cdt/commit/0a04a549c8910a5c4e8afa2b5f05178ddfe15d48#diff-09994977a11b6ae889c0499f59702255b3116a529d6526a687c7c21a1825a05cR106-R110) were accepted, it would clean up the code a bit and make some new things possible. That's for future consideration, though. 

At present, the most notable API change is the elimination of the mysterious `BLUEGRASS_HOM_META` macro which, I believe, was not functional anyways. But I'm still not certain what its purpose was...

An API addition, of sorts, is the inclusion of my TypeList template to make manipulating aggregates and variads of compile-time records easy, but that doesn't alter any existing API.

## Documentation Additions
- No Documentation Additions

### CODE REVIEW

**Comment by dimas1185** [comment]
hi Nate,
I don't know what was initial idea of bluegrass library. My latest changes were to use it as some lightweight `boost` replacement so we can deprecate `boost` in `multi_index` and other places. Also I'm not sure this is good candidate for broad usage in contracts.
On top of this you did two things:
you added your class `TypeList` to replace `std::tuple` inside bluegrass library.
on some reason you removed part of unit tests. Probably because of some functionality doesn't work now.

With all the respect to your work I do not understand why do we need those changes. You mentioned that your goal was to make Leap's and cdt's reflections compatible. That is legit point, we use different sets of classes in both. How this fix helps you to achieve that? If that makes sense we can connect on telegram and discuss.

**Comment by nathanielhourt** [comment]
OK, so let me take a step back for a moment, and point out why I'm bothering with this at all.

Hi, I'm Nate, and I have been working on the technology underlying Leap/EOSIO since before EOS was thought of. I was contracted to get the EOSIO project started back [before it was publicly announced](https://github.com/AntelopeIO/leap/commits/main?after=167baef13745e6f2e7df863249e8e5f701657708+22000&branch=main&qualified_name=refs%2Fheads%2Fmain), and I've been working on the reflection used by Leap since [long before that](https://github.com/bitshares/bitshares-fc/commit/72088c548be873e5921edb82ed516a0b37ca378e). I'm good at it by now, and I like it. =)

So why do we need these changes... well, the CDT's reflection right now (i.e., bluegrass's refl header) is a proof-of-concept level of implementation quality, but reflection is pretty straightforward so it's not a huge task to spiff it up and make it nice to use and build upon in the future. That's what this PReq does.

Going forward, I'd like to continue making updates to the reflection (already planned and implemented is adding support for reflecting inheritance hierarchies, but I want to get this in before making notable changes to the API) and I can also help out with any issues or feature requests that arise with respect to it in the future.

In short, I want to help make the CDT better.
Alas, though, I am something of a conscientious objector to Telegram and regard its use in the organization of open source projects as an antipattern, so I can't be found there, but feel free to reach out to my via email (you can pull my address out of my commits) and we can find a more convenient way to coordinate.

**Comment by nathanielhourt** [comment]
I'd like to understand further what your plans for bluegrass are, as you say you aren't sure it's good for broad usage in contracts... As `multi_index` is now based on it, literally all stateful contracts (i.e., all contracts) are utterly dependent upon it and its reflection infrastructure. Making that infrastructure as robust and powerful as possible, then, seems like a good idea. I also note that reflection is a more elegant, powerful, and accurate solution than the old PFR thing, and I applaud the removal of PFR.

Now, as for me, I can work with or without this patch upstream, but I'd really like to get it in because it cleans up the reflection to the point that I can make my system interoperable with it, and I want interoperability with the CDT's native logic. Without this patch, I can still implement my library over the CDT without interoperating with it, but it's better to play nicely. =)

I need to know, however, do I push ahead with or without interoperability with the CDT? In other words, do you and/or the CDT developers at large want my help? I'm happy to support the CDT's reflection and make it as powerful as Leap's (and more so -- Leap's is getting a bit old) but I can only help if my PReqs are welcome.

So once again, we want a better way to organize than just using GitHub, but given as Telegram continues to look and act like a honeypot, I'm just not going to work there. I could manage on-prem infrastructure for the CDT devs to coordinate if that is desirable. Hopefully soon I'll be able to offer hosting too, but right now that's beyond me.

Anyways, like I said, right now I need to know which way to go. Do you want my help with reflection, as it is now critical infrastructure for the CDT, and it really needs work? If so, what does it take to get this patch reviewed? Should we set up a call to go over the changes?

**Comment by dimas1185** [comment]
@nathanielhourt sorry for a delay. Currently we do not have any plans on bluegrass library updates except maybe bugfixes. It is also not considered to be part of contract developers tool set. However, I think we can communicate out of github to know more about your work and maybe reconsider bluegrass library usage. I need to wait someone to come from vacation to discuss.
I'll drop you an email to the email address you have on your profile. This will be next week or so.

**Comment by dimas1185** [comment]
currently `bluegrass` library remains internal cdt library and we do not advertise it for public usage. Library purpose is to serve `cdt` tool set since we want fixed code due to deterministic nature of blockchain. We are deprecating `boost` of `cdt` and `bluegrass` takes over that boost functionality. It can't compete with same purpose libraries like `boost::preprocessor` or `boost::hana` however it serves well taking our small scope.
If we are going to refactor it, I think next step would be to make it separate submodule and share with `eos-vm`. `eos-vm` has its own copy of that library that diverted a bit already. Also I'm not fond of the idea of touching what is smoothly working for years. Despite its small size and not being a good code example, we haven't had any bugs related except one that was introduced after boost library removal from `multi_index`.
So based on that I think this will be put aside until better times. You are welcome to give more justification to your changes, replacing standard library structures in favor of custom library must be very well justified. As you mentioned, this library is used by `multi_index` and virtually every single contract depends on it.
Anyway I excited about your effort to make `cdt` better and I know you have a good code knowledge based on your past PRs.

p.s. you also should hear from Eric via email.

**Comment by greg7mdp** [comment]
@nathanielhourt just curious, how would you compare your TypeList class with [boost mp11](https://github.com/boostorg/mp11)'s `mp_list`?

**Comment by nathanielhourt** [comment]
Similar, no doubt. Like Bluegrass, mine is more compact and is more directly applicable to the problems I am dealing with. :-)
